### DISCUSSION TITLE
syncing auxvariable and material computation

### DISCUSSION THREAD

**Discussion by mcacace** [idaholab/moose#20250] [Q&A General] [answered]
Dear all,
I would need to retrieve the old and current value of an auxvariable (computed at each linear iteration) inside a material. Is there a way to do so? 
Thanks a lot,
Mauro

**Comment by ajacquey** [comment]
Hi Mauro,

Did you try adding `execute_on = 'LINEAR'` in your `AuxKernel` block?
Retrieving the current value and old values of the `AuxVariable` in the material can then be done in the normal way with `coupledValue` and `coupledValueOld` I suppose.

Cheers,
Antoine

**Reply by mcacace** [reply]
Thanks @ajacquey (long time not seeing you). Did try to force the computation at linear (see my original post), but to not avail. I ended up by creating a copy of it via a stateful property. Admit it is not the most elegant solution, but it provides with more freedom for my case at least - I can restrict the actual computation to the end of the time step. 
Ciao,
Mauo

**Comment by lindsayad** [comment, marked as answer]
@mcacace material properties are indeed computed before auxiliary variables, e.g. code below, so you do have a dependency issue here. If you didn't want to use a stateful material property, you could use a material property for the current value and then an aux variable to get the old values (using `MaterialRealAux` as the mechanism to create the old aux values from the current material property). But it sounds like you are content with your work-around, which is great!
```c++
template <typename AuxKernelType>
void
ComputeElemAuxVarsThread<AuxKernelType>::onElement(const Elem * elem)
{
  if (_aux_kernels.hasActiveBlockObjects(_subdomain, _tid))
  {
    const std::vector<std::shared_ptr<AuxKernelType>> & kernels =
        _aux_kernels.getActiveBlockObjects(_subdomain, _tid);
    _fe_problem.prepare(elem, _tid);
    _fe_problem.reinitElem(elem, _tid);

    // Set up the sentinel so that, even if reinitMaterials() throws, we
    // still remember to swap back.
    SwapBackSentinel sentinel(_fe_problem, &FEProblem::swapBackMaterials, _tid, _need_materials);

    if (_need_materials)
      _fe_problem.reinitMaterials(elem->subdomain_id(), _tid);

    for (const auto & aux : kernels)
      aux->compute();
```
